docs/transaction-api.md

```diff
@@ -60,6 +60,53 @@ A payment in the `Cancelled` state may be explicitly retried ("requoted") by the
 - A `FixedSend` payment will attempt to pay `intent.amountToSend - amountAlreadySent`.
 - A `FixedDelivery` payment will attempt to pay the remaining `invoice.amount - invoice.received` (according to the remote invoice state).
 
+## Incoming Payment Lifecycle
+
+### Invoice creation
+
+An invoice is created according to the [Open Payments](https://docs.openpayments.dev/invoices#create) specification. Rafiki creates a payment account for each invoice.
+
+### Receiving
+
+An invoice receives funds via Interledger as long as it is active.
+
+### Deactivation
+
+An invoice is deactivated when either:
+
+- it has received its specified `amount`
+- it has expired
+
+When the invoice is deactivated, Rafiki notifies the wallet of received funds via `invoice.paid` or `invoice.expired` [webhook events](#webhooks).
+
+An expired invoice that has never received money is deleted.
+
+## Webhooks
+
+Rafiki sends webhook events to notify the wallet of payment lifecycle states that require liquidity to be added or removed.
+
+The `outgoing_payment.funding` event should be handled idempotently. The `Rafiki-Signature` header may be used as an idempotency key, or the wallet may call `Query.getOutgoingPayment` to retrieve `quote.maxSourceAmount` before calling `Mutation.fundOutgoingPayment` with that `amount`.
```

Comment:
Don't all the webhooks need to handled idempotently (if that's a word)?

e.g. `handleCancelled` if the transaction aborted, the caller will retry it, possible re-triggering the webhook.

Reply:
:thinking: I was mainly focusing on the liquidity calls.
Does `Mutation.requoteOutgoingPayment` being idempotent make that ok? You can only retry a payment that's currently `Cancelled`

Reply:
I think we might be getting our wires crossed.

To elaborate my example, consider this sequence of events:
1. `cancelPayment` is called (on a payment with state `Funding`),
2. (begin transaction)
3. `handleCancelled` is called...
4. update `payment.state` to `Cancelled`
5. Send a `PaymentCancelled` webhook.
6. (try to commit the transaction)
7. Commit failed, because someone crashed or a connection dropped (etc)
8. Oh well, the client retries `cancelPayment`.
9. Everything is fine, but the webhook is called a second time.

So unless I'm missing something the web hook either needs to be idempotent or have no side effects.

I haven't checked every webhook request, but I assume they have similar constraints.